src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,674 @@
+        all_attributes = [attribute for attribute in cluster.Attributes.__dict__.values() if inspect.isclass(
+            attribute) and issubclass(attribute, ClusterObjects.ClusterAttributeDescriptor)]
+
+        # Hackish way to get enums to return properly -- the default behavior (under else block) returns a BLANK LIST without this workaround
+        # If type(attribute.attribute_type.Type) or type(ClusterObjects.ClusterObjectFieldDescriptor(Type=desired_type).Type are enums, they return <class 'aenum._enum.EnumType'>, which are equal!
+        if desired_type == MatterIntEnum:
+            all_attributes_of_type = [attribute for attribute in all_attributes if type(
+                attribute.attribute_type.Type) == type(ClusterObjects.ClusterObjectFieldDescriptor(Type=desired_type).Type)]
+
+        elif desired_type == IntFlag:
+            if hasattr(cluster, 'Attributes'):
+                attributes_class = getattr(cluster, 'Attributes')
+                if hasattr(attributes_class, 'FeatureMap'):
+                    all_attributes_of_type = [cluster.Attributes.FeatureMap]
+                else:
+                    raise Exception(f'Cluster {cluster} lacks a FeatureMap')
+            else:
+                raise Exception(f'Cluster {cluster} lacks attributes')
+        else:
+            all_attributes_of_type = [attribute for attribute in all_attributes if attribute.attribute_type ==
+                                      ClusterObjects.ClusterObjectFieldDescriptor(Type=desired_type)]
+
+        return all_attributes_of_type
+
+    def all_device_clusters(self) -> set:
+        device_clusters = set()
+        for endpoint in self.endpoints:
+            device_clusters |= set(self.endpoints[endpoint].keys())
+        return device_clusters
+
+    def all_device_attributes(self) -> set:
+        device_attributes = set()
+        for endpoint in self.endpoints:
+            for cluster in self.endpoints[endpoint]:
+                device_attributes |= self.endpoints[endpoint][cluster].keys()
+        return device_attributes
+
+    def verify_attribute_list_cluster(self, read_request: dict, endpoint: int, cluster: ClusterObjects.Cluster):
+        cluster_ids = [c.id for c in read_request[endpoint].keys()]
+        asserts.assert_equal(
+            sorted(read_request[endpoint][Clusters.Descriptor][Clusters.Descriptor.Attributes.ServerList]),
+            sorted(cluster_ids),
+            f"ServerList doesn't match the expected server list for endpoint {endpoint}"
+        )
+
+    async def check_attribute_read_for_type(self, desired_attribute_type: type) -> bool:
+        # Get all clusters from device
+        for cluster in self.device_clusters:
+            all_types = await self.all_type_attributes_for_cluster(cluster, desired_attribute_type)
+            all_types = list(set(all_types) & self.device_attributes)
```

Comment:
This isn't going to get you what you want - this list is just a list of all the device attribute ids for every cluster all smooshed together. This needs to be across clusters.

Ex. limiting this to just one endpoint for now
say you have this
Cluster 1 defines attributes 1, 2 and 3. The device has 2 and 3. 
Cluster 2 defines attributes 1, 2, 3 and 4. The device supports 1 and 4

Per the code in all_device_attributes, self.device_attributes will contain {1,2,3,4}. But your dut doesn't support attribute 1 for cluster 1, and that's would be selected here.